### ISSUE TITLE
Enterprise Gateway on k8s spawning YARN Cluster kernels

### ISSUE THREAD

**Issue by jpugliesi** [jupyter/enterprise_gateway#814]
## Description
I am attempting to configure the following end to end architecture:
* Deploy a single Enterprise Gateway server on Kubernetes with a `LoadBalancer` `Service` (internal to the AWS private subnet)
* Deploy multiple AWS EMR clusters (running YARN/Spark) as the targets for kernel deployment
* Configure the Enterprise Gateway with, for example, one kernelspec per EMR cluster (configuring the relevant EG_YARN_ENDPOINT, etc) to do so.

In theory, a benefit of this approach, among others, is that there'd only be one EG deployment to operate (as opposed to one EG running on each EMR cluster master node).

So far, I am able to begin launching a JEG kernel in a YARN container on a remote EMR cluster (via yarn cluster mode). Problem is, the kernel on YARN hangs when it attempts to send its connection info back to the EG (from the YARN logs):
```
[D 2020-05-07 17:54:55,845.845 launch_ipykernel] Using connection file '/tmp/kernel-2b40c038-3124-414e-9b7d-bf2b486d67a7_q0feo9ui.json'.
[I 2020-05-07 17:54:55,846.846 launch_ipykernel] Signal socket bound to host: 0.0.0.0, port: 55867
...... # nothing
```

As far as I can tell, this is a network issue - the kernel cant network with the EG's `{response_address}`. The EG is running as a `LoadBalancer` Service on K8s, but the response address it generates is not accessible outside of the k8s cluster. Specifically, the EG generates a _dynamic_ port from the `EG_PORT_RANGE`, but that `{load_balancer_ip}:{generated_port}` combination wont be accessible unless the `LoadBalancer` Service explicitly defines  all possible ports in the `EG_PORT_RANGE`

A couple questions:
1. Am I understanding the behavior correctly?
2. Is this approach, running EG on k8s to spawn Spark kernels on YARN (cluster mode) fundamentally a bad idea, or is anyone aware of alternative approaches?
3. JEG requires 6 ports/connections with each kernel - I'm curious - how does the KubernetesProcessProxy define these 6 connections between the EG pod, and a kernel pod? (I've browsed the source, but didnt find any indication of 6 ports being opened in the k8s launch template.

As always, appreciate the help.

## Screenshots / Logs
JEG logs upon kernel start (`spark-submit`):
```
Starting IPython kernel for Spark in Yarn Cluster mode as user hadoop

++ exec /usr/hdp/current/spark2-client/bin/spark-submit --master yarn --deploy-mode cluster --name 2b40c038-3124-414e-9b7d-bf2b486d67a7 --packages com.amazonaws:aws-java-sdk:1.7.4,org.apache.hadoop:hadoop-aws:2.7.7 --conf spark.yarn.dist.archives=s3a://some-s3-bucket/pyspark.tar.gz#environment --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.appMasterEnv.PYTHONPATH=./environment/bin/ --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=./environment/bin/python --conf spark.yarn.executorEnv.PYSPARK_PYTHON=./environment/bin/python --conf spark.yarn.appMasterEnv.PATH=./environment/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin --proxy-user hadoop /usr/local/share/jupyter/kernels/spark_python_yarn_cluster/scripts/launch_ipykernel.py --RemoteProcessProxy.kernel-id 2b40c038-3124-414e-9b7d-bf2b486d67a7 --RemoteProcessProxy.response-address 10.100.147.149:47397 --RemoteProcessProxy.port-range 0..0 --RemoteProcessProxy.spark-context-initialization-mode lazy

```

JEG logs (after successfully scheduling the kernel on YARN:
```
...
[D 2020-05-07 17:58:32.179 EnterpriseGatewayApp] Waiting for KernelID '2b40c038-3124-414e-9b7d-bf2b486d67a7' to send connection info from host 'ip-10-0-228-187.us-west-2.compute.internal' - retrying...
[D 2020-05-07 17:58:32.683 EnterpriseGatewayApp] 121: State: 'ACCEPTED', Host: 'ip-10-0-228-187.us-west-2.compute.internal', KernelID: '2b40c038-3124-414e-9b7d-bf2b486d67a7', ApplicationID: 'application_1588864343234_0014'
[D 2020-05-07 17:58:37.686 EnterpriseGatewayApp] Waiting for KernelID '2b40c038-3124-414e-9b7d-bf2b486d67a7' to send connection info from host 'ip-10-0-228-187.us-west-2.compute.internal' - retrying...
[D 2020-05-07 17:58:38.190 EnterpriseGatewayApp] 122: State: 'ACCEPTED', Host: 'ip-10-0-228-187.us-west-2.compute.internal', KernelID: '2b40c038-3124-414e-9b7d-bf2b486d67a7', ApplicationID: 'application_1588864343234_0014'
[D 2020-05-07 17:58:43.194 EnterpriseGatewayApp] Waiting for KernelID '2b40c038-3124-414e-9b7d-bf2b486d67a7' to send connection info from host 'ip-10-0-228-187.us-west-2.compute.internal' - retrying...
[D 2020-05-07 17:58:43.697 EnterpriseGatewayApp] 123: State: 'ACCEPTED', Host: 'ip-10-0-228-187.us-west-2.compute.internal', KernelID: '2b40c038-3124-414e-9b7d-bf2b486d67a7', ApplicationID: 'application_1588864343234_0014'
[D 2020-05-07 17:58:48.702 EnterpriseGatewayApp] Waiting for KernelID '2b40c038-3124-414e-9b7d-bf2b486d67a7' to send connection info from host 'ip-10-0-228-187.us-west-2.compute.internal' - retrying...
[W 2020-05-07 17:58:49.208 EnterpriseGatewayApp] Termination of application 'application_1588864343234_0014' failed with exception: 'Response finished with status: 401. Details: {"RemoteException":{"exception":"AuthorizationException","message":"Unable to obtain user name, user not authenticated","javaClassName":"org.apache.hadoop.security.authorize.AuthorizationException"}}'.  Continuing...
[D 2020-05-07 17:58:49.211 EnterpriseGatewayApp] YarnClusterProcessProxy.kill, application ID: application_1588864343234_0014, kernel ID: 2b40c038-3124-414e-9b7d-bf2b486d67a7, state: FAILED, result: None
[E 2020-05-07 17:58:49.211 EnterpriseGatewayApp] KernelID: '2b40c038-3124-414e-9b7d-bf2b486d67a7' launch timeout due to: YARN resources unavailable after 303.0 seconds for app application_1588864343234_0014, launch timeout: 300.0!  Check YARN configuration.
```

YARN kernel container logs (stdout) - notice that it hangs after the `Signal socket bound to host` line:
```
[D 2020-05-07 17:54:55,845.845 launch_ipykernel] Using connection file '/tmp/kernel-2b40c038-3124-414e-9b7d-bf2b486d67a7_q0feo9ui.json'.
[I 2020-05-07 17:54:55,846.846 launch_ipykernel] Signal socket bound to host: 0.0.0.0, port: 55867
Heap
 garbage-first heap   total 249856K, used 107519K [0x00000006c0000000, 0x00000006c01007a0, 0x00000007c0000000)
  region size 1024K, 106 young (108544K), 9 survivors (9216K)
 Metaspace       used 27786K, capacity 28004K, committed 28288K, reserved 1073152K
  class space    used 3501K, capacity 3594K, committed 3712K, reserved 1048576K
 Concurrent marking:
      0   init marks: total time =     0.00 s (avg =     0.00 ms).
      1      remarks: total time =     0.00 s (avg =     3.43 ms).
           [std. dev =     0.00 ms, max =     3.43 ms]
         1  final marks: total time =     0.00 s (avg =     0.13 ms).
              [std. dev =     0.00 ms, max =     0.13 ms]
         1    weak refs: total time =     0.00 s (avg =     3.30 ms).
              [std. dev =     0.00 ms, max =     3.30 ms]
      1     cleanups: total time =     0.00 s (avg =     0.40 ms).
           [std. dev =     0.00 ms, max =     0.40 ms]
    Final counting total time =     0.00 s (avg =     0.10 ms).
    RS scrub total time =     0.00 s (avg =     0.14 ms).
  Total stop_world time =     0.00 s.
  Total concurrent time =     0.00 s (    0.00 s marking).
```

YARN kernel container logs (stderr) - times out:
```
20/05/07 17:54:54 WARN conf.Configuration: __spark_hadoop_conf__.xml:an attempt to override final parameter: yarn.nodemanager.local-dirs;  Ignoring.
20/05/07 17:54:54 INFO yarn.ApplicationMaster: Preparing Local resources
20/05/07 17:54:55 WARN conf.Configuration: __spark_hadoop_conf__.xml:an attempt to override final parameter: fs.s3.buffer.dir;  Ignoring.
20/05/07 17:54:55 WARN conf.Configuration: __spark_hadoop_conf__.xml:an attempt to override final parameter: yarn.nodemanager.local-dirs;  Ignoring.
20/05/07 17:54:55 INFO yarn.ApplicationMaster: ApplicationAttemptId: appattempt_1588864343234_0014_000001
20/05/07 17:54:55 INFO yarn.ApplicationMaster: Starting the user application in a separate Thread
20/05/07 17:54:55 INFO yarn.ApplicationMaster: Waiting for spark context initialization...
20/05/07 17:56:35 ERROR yarn.ApplicationMaster: Uncaught exception: 
java.util.concurrent.TimeoutException: Futures timed out after [100000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:223)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:227)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:220)
	at org.apache.spark.deploy.yarn.ApplicationMaster.runDriver(ApplicationMaster.scala:468)
	at org.apache.spark.deploy.yarn.ApplicationMaster.org$apache$spark$deploy$yarn$ApplicationMaster$$runImpl(ApplicationMaster.scala:305)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply$mcV$sp(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$3.run(ApplicationMaster.scala:779)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1698)
	at org.apache.spark.deploy.yarn.ApplicationMaster.doAsUser(ApplicationMaster.scala:778)
	at org.apache.spark.deploy.yarn.ApplicationMaster.run(ApplicationMaster.scala:244)
	at org.apache.spark.deploy.yarn.ApplicationMaster$.main(ApplicationMaster.scala:803)
	at org.apache.spark.deploy.yarn.ApplicationMaster.main(ApplicationMaster.scala)
20/05/07 17:56:35 INFO yarn.ApplicationMaster: Final app status: FAILED, exitCode: 13, (reason: Uncaught exception: java.util.concurrent.TimeoutException: Futures timed out after [100000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:223)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:227)
	at org.apache.spark.util.ThreadUtils$.awaitResult(ThreadUtils.scala:220)
	at org.apache.spark.deploy.yarn.ApplicationMaster.runDriver(ApplicationMaster.scala:468)
	at org.apache.spark.deploy.yarn.ApplicationMaster.org$apache$spark$deploy$yarn$ApplicationMaster$$runImpl(ApplicationMaster.scala:305)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply$mcV$sp(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anonfun$run$1.apply(ApplicationMaster.scala:245)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$3.run(ApplicationMaster.scala:779)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1698)
	at org.apache.spark.deploy.yarn.ApplicationMaster.doAsUser(ApplicationMaster.scala:778)
	at org.apache.spark.deploy.yarn.ApplicationMaster.run(ApplicationMaster.scala:244)
	at org.apache.spark.deploy.yarn.ApplicationMaster$.main(ApplicationMaster.scala:803)
	at org.apache.spark.deploy.yarn.ApplicationMaster.main(ApplicationMaster.scala)
)
20/05/07 17:56:35 INFO util.ShutdownHookManager: Shutdown hook called
```

kernelspec:
```
{
  "language": "python",
  "display_name": "Spark - Python (YARN Cluster Mode)",
  "metadata": {
    "process_proxy": {
      "class_name": "enterprise_gateway.services.processproxies.yarn.YarnClusterProcessProxy"
    }
  },
  "env": {
    "SPARK_HOME": "/usr/hdp/current/spark2-client",
    "PYSPARK_PYTHON": "/opt/conda/bin/python",
    "HADOOP_USER_NAME": "hadoop",
    "YARN_CONF_DIR": "/etc/yarn/conf/",
    "SPARK_CONF_DIR": "/etc/spark/conf/",
    "HADOOP_CONF_DIR": "/etc/hadoop/conf/",
    "EG_YARN_ENDPOINT": "${EG_YARN_ENDPOINT}",
    "EG_IMPERSONATION_ENABLED": "True",
    "PYTHONPATH": "${HOME}/.local/lib/python3.7/site-packages:/usr/hdp/current/spark2-client/python:/usr/hdp/current/spark2-client/python/lib/py4j-0.10.6-src.zip",

    "SPARK_OPTS": "--master yarn --deploy-mode cluster --name ${KERNEL_ID:-ERROR__NO__KERNEL_ID} --packages com.amazonaws:aws-java-sdk:1.7.4,org.apache.hadoop:hadoop-aws:2.7.7 --conf spark.yarn.dist.archives=s3a://some-s3-bucket/pyspark.tar.gz#environment --conf spark.yarn.submit.waitAppCompletion=false --conf spark.yarn.appMasterEnv.PYTHONPATH=./environment/bin/ --conf spark.yarn.appMasterEnv.PYSPARK_PYTHON=./environment/bin/python --conf spark.yarn.executorEnv.PYSPARK_PYTHON=./environment/bin/python --conf spark.yarn.appMasterEnv.PATH=./environment/bin:$PATH ${KERNEL_EXTRA_SPARK_OPTS}",
    "LAUNCH_OPTS": ""
  },
  "argv": [
    "/usr/local/share/jupyter/kernels/spark_python_yarn_cluster/bin/run.sh",
    "--RemoteProcessProxy.kernel-id",
    "{kernel_id}",
    "--RemoteProcessProxy.response-address",
    "{response_address}",
    "--RemoteProcessProxy.port-range",
    "{port_range}",
    "--RemoteProcessProxy.spark-context-initialization-mode",
    "lazy"
  ]
}
```
## Environment

- Enterprise Gateway Version [e.g. 1.x, 2.x, ...]
- Platform: [e.g. YARN, Kubernetes ...]
- Others [e.g. Jupyter Server 5.7, JupyterHub 1.0, etc]

**Comment by kevin-bates** [comment]
>1. Am I understanding the behavior correctly?

Yes, the issue you are experiencing is expected today.  EG is (currently) meant to service one type of cluster and be on the same network as those kernels.  The issue you're having is that the response address that EG includes in the launch of kernels across the YARN cluster is, as far as the YARN nodes are concerned, invalid.  

I think this would be great functionality to support, but it doesn't at the moment and this would need to be an external contribution at this time - sorry.

>2. Is this approach, running EG on k8s to spawn Spark kernels on YARN (cluster mode) fundamentally a bad idea, or is anyone aware of alternative approaches?

What is the reason you need to run EG in k8s?  Is it merely to benefit from the replication set functionality?  You should be able to hit different YARN RMs by pushing the configuration of the [YARN endpoint](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/config-options.html#per-kernel-configuration-overrides) into each of the [kernelspec files](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/system-architecture.html#process-proxy-configuration).  Any configuration settings at that level override the global settings.

Also note that any use of load-balancers, inside or out of k8s, require session-affinity.

>3. JEG requires 6 ports/connections with each kernel - I'm curious - how does the KubernetesProcessProxy define these 6 connections between the EG pod, and a kernel pod? (I've browsed the source, but didnt find any indication of 6 ports being opened in the k8s launch template.

This is one reason why EG and kernels need to be on the same network.  We don't expose the 6 ports in the container launch, just using the container network.  As a result, this too might be an issue unless you can configure host network functionality in your k8s instances.

I'd also love to remove the need for EG to have to run in k8s to launch kernel pods, but then you are talking about exposure of the ports.  What would be really nice is to push the WebSocket handling down into the kernel pod, then we'd only need to publish a single port and we could probably manage that.

**Comment by jpugliesi** [comment]
@kevin-bates thanks as always for a thorough response. This all makes sense

> What is the reason you need to run EG in k8s? Is it merely to benefit from the replication set functionality? You should be able to hit different YARN RMs by pushing the configuration of the YARN endpoint into each of the kernelspec files. Any configuration settings at that level override the global settings.

We'd like to run EG in k8s for a variety of reasons, but the main ones being:
* k8s `Deployment`s make it easy to create a reliable EG 
* Our team is most familiar with managing k8s workloads, so keeping the infrastructure consistent simplifies things from an operational perspective
* We're managing a multi-tenant system, where each tenant's data needs to be processed and stored in isolation (i.e. Spark jobs need to run in fully separate Spark/YARN clusters). The idea would be to run a "hub and spoke" architecture, where a central k8s cluster manages multiple EG deployments. Each EG deployment then connects to its own YARN RM(s). Each YARN RM would manage its own isolated YARN cluster, allowing us to run notebooks in separate tenant environments, but managing the EG instances in a single cluster/location.

I'm aware of EG's ability to connect to multiple HA YARN RMs - but thats not really our goal here.

> I'd also love to remove the need for EG to have to run in k8s to launch kernel pods, but then you are talking about exposure of the ports. What would be really nice is to push the WebSocket handling down into the kernel pod, then we'd only need to publish a single port and we could probably manage that.

I was wondering if it would be possible to push the websocket handling down into the kernel pod - that sounds like a pretty serious jupyter api change, but I'm not sure. Do you have any resources/pointers I could explore to understand what it would take to develop this?

**Comment by kevin-bates** [comment]
> I was wondering if it would be possible to push the websocket handling down into the kernel pod - that sounds like a pretty serious jupyter api change, but I'm not sure. Do you have any resources/pointers I could explore to understand what it would take to develop this?

I'm not aware of any.  The problem is that the WebSocket is separated into the ZMQ "channels" via [`ZMQChannelsHandler`](https://github.com/jupyter/notebook/blob/master/notebook/services/kernels/handlers.py#L93) and that is engrained in Notebook.  So, EG would have to subclass that handler, check if the kernel supports a "proxied WebSocket" and, if so, proxy it further; else fallback to today's behavior.

If proxied, the target would essentially have the ZMQChannelsHandler and jupyter_client code with a local kernel.  I would think we'd simply extend our kernel-launchers with that kind of code.

But that is more for running EG outside of k8s, where we could expose a single port into the kernel pod - rather than 6.  

However, it just dawned on me that what you might have more of an issue with are the different EG response addresses that the YARN-based kernels need to respond back to.  This is kind of the opposite problem (but probably easier to resolve) and I apologize for not having realized this sooner.

We currently create a different response socket for each kernel launched.  As a result, there's no way for you to "provision" the k8s container to expose N-many ports for response addresses.  I've thought that we could probably go down to a single response port where each responder encodes their kernel-id in the payload and a thread or some ioloop thing (sorry, I'm not up on that terminology) doles out responses to the appropriate kernel manager instances.  That way you could say all responses go to port 8899 (for example), allowing the k8s EG service to expose that port, in addition to the traditional front-end 8888 port.  The IP for the port would need to be an IP the YARN clusters can hit as well, but that's probably the result of the LB config.

Does the _response-address issue_ sound more like what you're issue really is?  You might be able to experiment with pinning a port (see `_prepare_response_socket`):
https://github.com/jupyter/enterprise_gateway/blob/dae118e34185f04bbb2e8124fd2419008cb40add/enterprise_gateway/services/processproxies/processproxy.py#L714-L723
for all kernels, expose that port via the k8s service/LB and, so long as you don't overlap kernel startups, at least see if that is sufficient.  The port is ONLY necessary to receive the kernel's connection info, after which its discarded (so you probably want to comment out [that code](https://github.com/jupyter/enterprise_gateway/blob/dae118e34185f04bbb2e8124fd2419008cb40add/enterprise_gateway/services/processproxies/processproxy.py#L945-L951) as well) - for the experiment.  

Since EG kernel starts are still single-threaded (until NB 6.1 is available) you probably don't have to even worry about throttling your kernel starts - since the framework does it for you. :smile:

I hope this is all that you need.  I think it might actually.

**Comment by dnks23** [comment]
Hi!
I am trying kind of the same setup (Hub & EG on K8s spawing Yarn-Kernels on a single Yarn-Cluster). I wonder whether you @jpugliesi already tried out the above option from @kevin-bates ? Were you actually able to get this working? Or where ended you up?

**Comment by kevin-bates** [comment]
FWIW, I hope to build the single-response-address approach into EG 3.0 (or 3.1).

**Comment by kevin-bates** [comment]
This work in 3.0 might be of interest here: https://github.com/jupyter/enterprise_gateway/pull/877

**Comment by kevin-bates** [comment]
Let's see if #877 helps this out.  It will be available in the EG 2.6.0 release - which should be available soon.
